File: ui/src/main/java/com/peakmain/ui/dialog/AlertController.java
package com.peakmain.ui.dialog;

import android.content.Context;
import android.content.DialogInterface;
import android.util.SparseArray;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;

/**
 * author peakmain
 * createdata：2019/7/9
 * mail:2726449200@qq.com
 * desiption:Special placement parameters
 */
public class AlertController {

    private Window mWindow;
    private AlertDialog mAlertDialog;
    private ViewHelper mViewHelper;

    public AlertController(AlertDialog dialog, Window window) {
        this.mWindow = window;
        this.mAlertDialog = dialog;

    }

    public Window getWindow() {
        return mWindow;
    }

    public AlertDialog getDialog() {
        return mAlertDialog;
    }

    public void setText(int viewId, CharSequence text) {
        mViewHelper.setText(viewId, text);
    }

    public void setViewHelper(ViewHelper mViewHelper) {
        this.mViewHelper = mViewHelper;
    }


    public void setOnClickListener(int viewId, View.OnClickListener listener) {
        mViewHelper.setOnClickListener(viewId, listener);
    }

    public <T extends View> T getView(int viewId) {
        return mViewHelper.getView(viewId);
    }

    static class AlertParams {
        public Context mContext;
        public int themeResId;
        //click blank to cancel
        public boolean mCancelable = true;
        //dialog Cancel
        public DialogInterface.OnCancelListener mOnCancelListener;
        //dialog Dismiss
        public DialogInterface.OnDismissListener mOnDismissListener;
        //Keys to monitor
        public DialogInterface.OnKeyListener mOnKeyListener;
        //Loading layout  view
        public View mView;
        //id  loading layout
        public int mViewLayoutResId;
        //save the text in the custom view
        SparseArray<CharSequence> mTextArray = new SparseArray<>();
        ///save the button click in the custom view
        SparseArray<View.OnClickListener> mClickArray = new SparseArray<>();
        public int mWidth = ViewGroup.LayoutParams.WRAP_CONTENT;
        public int mGravity = Gravity.CENTER;
        public int mAnimation;
        public int mHeigth = ViewGroup.LayoutParams.WRAP_CONTENT;

        public AlertParams(Context context, int themeResId) {
            this.mContext = context;
            this.themeResId = themeResId;
        }

        /**
         * Bind and set parameters
         */
        public void apply(AlertController mAlert) {
            //set parameters
            //1.set layout
            ViewHelper viewHelper = null;
            if (mViewLayoutResId != 0) {
                viewHelper = new ViewHelper(mContext, mViewLayoutResId);
            }
            if (mView != null) {
                viewHelper = new ViewHelper();
                viewHelper.setContentView(mView);
            }
            if (viewHelper == null) {
                throw new IllegalArgumentException("Plase Set the layout setContentView()");
            }
            //Set the layout for the dialog
            mAlert.mAlertDialog.setContentView(viewHelper.getContentView());
            //2.set text
            int textArraysize = mTextArray.size();
            for (int i = 0; i < textArraysize; i++) {
                viewHelper.setText(mTextArray.keyAt(i), mTextArray.valueAt(i));
            }
            //3.set click
            int clickSize = mClickArray.size();
            for (int i = 0; i < clickSize; i++) {
                viewHelper.setOnClickListener(mClickArray.keyAt(i), mClickArray.valueAt(i));
            }
            mAlert.setViewHelper(viewHelper);
            //4.Configure custom effects to pop up the default animation from the bottom of the full screen
            Window window = mAlert.getWindow();
            //set gravity
            window.setGravity(mGravity);
            if (mAnimation != 0) {
                window.setWindowAnimations(mAnimation);
            }
            //set width and height
            WindowManager.LayoutParams layoutParams = window.getAttributes();
            layoutParams.height = mHeigth;
            layoutParams.width = mWidth;
            window.setAttributes(layoutParams);
        }


    }
}

File: ui/src/main/java/com/peakmain/ui/dialog/AlertController.kt
package com.peakmain.ui.dialog

import android.content.Context
import android.content.DialogInterface
import android.os.Build.VERSION_CODES.P
import android.util.SparseArray
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.view.WindowManager
import com.peakmain.ui.utils.SizeUtils.screenHeight

/**
 * author peakmain
 * create：2019/7/9
 * mail:2726449200@qq.com
 * description:Special placement parameters
 */
class AlertController(val dialog: AlertDialog, val window: Window) {
    private lateinit var mViewHelper: ViewHelper

    fun setText(viewId: Int, text: CharSequence?) {
        mViewHelper.setText(viewId, text)
    }

    fun setOnClickListener(viewId: Int, listener: View.OnClickListener?) {
        mViewHelper.setOnClickListener(viewId, listener)
    }

    fun <T : View> getView(viewId: Int): T? {
        return mViewHelper.getView(viewId)
    }

    internal class AlertParams(var mContext: Context, var themeResId: Int) {


        //click blank to cancel
        @JvmField
        var mCancelable = true

        //dialog Cancel
        @JvmField
        var mOnCancelListener: DialogInterface.OnCancelListener? = null

        //dialog Dismiss
        @JvmField
        var mOnDismissListener: DialogInterface.OnDismissListener? = null

        //Keys to monitor
        @JvmField
        var mOnKeyListener: DialogInterface.OnKeyListener? = null

        //Loading layout  view
        @JvmField
        var mView: View? = null

        //id  loading layout
        @JvmField
        var mViewLayoutResId = 0

        //save the text in the custom view
        @JvmField
        var mTextArray = SparseArray<CharSequence>()

        ///save the button click in the custom view
        @JvmField
        var mClickArray = SparseArray<View.OnClickListener>()

        @JvmField
        var mWidth = ViewGroup.LayoutParams.WRAP_CONTENT

        @JvmField
        var mGravity = Gravity.CENTER

        @JvmField
        var mAnimation = 0

        @JvmField
        var mHeight = ViewGroup.LayoutParams.WRAP_CONTENT

        @JvmField
        var mMaxHeight: Double = 0.0


        /**
         * Bind and set parameters
         */
        fun apply(mAlert: AlertController) {
            //set parameters
            //1.set layout
            var viewHelper: ViewHelper? = null
            if (mViewLayoutResId != 0) {
                viewHelper = ViewHelper(mContext, mViewLayoutResId)
            }
            if (mView != null) {
                viewHelper = ViewHelper()
                viewHelper.contentView = mView
            }
            requireNotNull(viewHelper) { "Plase Set the layout setContentView()" }
            //Set the layout for the dialog
            mAlert.dialog.setContentView(viewHelper.contentView!!)
            //2.set text
            val textArraysize = mTextArray.size()
            for (i in 0 until textArraysize) {
                viewHelper.setText(mTextArray.keyAt(i), mTextArray.valueAt(i))
            }
            //3.set click
            val clickSize = mClickArray.size()
            for (i in 0 until clickSize) {
                viewHelper.setOnClickListener(mClickArray.keyAt(i), mClickArray.valueAt(i))
            }
            mAlert.mViewHelper = viewHelper
            //4.Configure custom effects to pop up the default animation from the bottom of the full screen
            val window = mAlert.window
            //set gravity
            window.setGravity(mGravity)
            if (mAnimation != 0) {
                window.setWindowAnimations(mAnimation)
            }
            //set width and height
            val layoutParams = window.attributes
            if(mMaxHeight>0){
                when (layoutParams.height) {
                    WindowManager.LayoutParams.MATCH_PARENT -> {
                        layoutParams.height = screenHeight.coerceAtMost(mMaxHeight.toInt());  // 如果是MATCH_PARENT，设置为屏幕高度和最大高度的较小值
                    }
                    WindowManager.LayoutParams.WRAP_CONTENT -> {
                        layoutParams.height = screenHeight.coerceAtMost(mMaxHeight.toInt()) // 可以加一个最大值限制
                    }
                    else -> {
                        // 如果不是MATCH_PARENT或WRAP_CONTENT，直接设置最大高度
                        layoutParams.height = layoutParams.height.coerceAtMost(mMaxHeight.toInt())
                    }
                }
            }else{
                layoutParams.height = mHeight
            }
            layoutParams.width = mWidth
            window.attributes = layoutParams
        }

    }

}